feat: mermaid diagram support in the chat

ConstantTime opened the pull request:
## Summary

Implements native inline Mermaid chart rendering in LibreChat chat interface, addressing feature request #7848. This enhancement replaces the current heavy artifacts-based approach with a lightweight, fast inline rendering system that displays Mermaid diagrams directly within chat messages.

**Key Benefits:**
- **Faster rendering**: Diagrams appear instantly inline without opening separate panels
- **Theme-aware**: Automatically adapts to LibreChat's dark/light themes  
- **Responsive design**: Optimized sizing for chat context with mobile support
- **Smart error handling**: Auto-correction for common syntax issues with helpful error messages
- **Performance optimized**: Debounced rendering prevents excessive re-renders during streaming

The implementation detects Mermaid code blocks in chat messages and renders them as inline SVG diagrams while maintaining full compatibility with existing code block functionality.

## Change Type

- [x] New feature (non-breaking change which adds functionality)

## Testing

**Test the Mermaid rendering functionality:**

In Dark mode

<img width="821" alt="Screenshot 2025-06-29 at 12 29 28 AM" src="https://github.com/user-attachments/assets/a88f9cd8-f26c-425d-b3fc-ecefc0215ac2" />

---

In Light mode

<img width="839" alt="Screenshot 2025-06-29 at 12 56 39 AM" src="https://github.com/user-attachments/assets/1a71d300-d343-4e26-944e-d8a39d616fa9" />

---

End to End recording

https://github.com/user-attachments/assets/a155b0c6-9640-4ef1-830e-4b592244227f

---

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [x] My changes do not introduce new warnings
- [x] Local unit tests pass with my changes
- [x] Feature tested across different diagram types and themes

danny-avila commented on client/src/components/Chat/Messages/Content/MermaidDiagram.tsx:
```diff
@@ -0,0 +1,327 @@
+    if (fixedContent !== content) {
+      // Currently just copies the fixed version to clipboard
+      navigator.clipboard.writeText(fixedContent).then(() => {
+        setError(`Potential fix copied to clipboard. Common issues found and corrected.`);
```
Non-localized text: translation should be added to `client/src/locales/en/translation.json` and this file, which is for English, only

danny-avila commented on client/src/components/Chat/Messages/Content/MermaidDiagram.tsx:
```diff
@@ -0,0 +1,327 @@
+              setWasAutoCorrected(true);
+            } catch (_fixedRenderError) {
+              if (!isCancelled) {
+                setError('Invalid diagram syntax - syntax errors found but unable to auto-correct');
```
Non-localized text: please add localization text (English only)

danny-avila commented on client/src/components/Chat/Messages/Content/MermaidDiagram.tsx:
```diff
@@ -0,0 +1,327 @@
+          }
+        } else {
+          if (!isCancelled) {
+            setError('No SVG generated - rendering failed unexpectedly');
```
Non-localized text: please add localization text (English only)

danny-avila commented:
Thanks @ConstantTime

I marked a few instances of text that require localization, but please double-check all user-facing text is localized.

danny-avila commented on client/vite.config.ts:
```diff
@@ -47,7 +47,7 @@ export default defineConfig(({ command }) => ({
         ],
         globIgnores: ['images/**/*', '**/*.map', 'index.html'],
-        maximumFileSizeToCacheInBytes: 4 * 1024 * 1024,
+        maximumFileSizeToCacheInBytes: 8 * 1024 * 1024,
```
does the mermaid library really add this much to the bundle size? Is there any way we can have a leaner implementation?

ConstantTime replied:
shifted to dynamic imports here to reduce the same.

ConstantTime commented on client/src/components/Chat/Messages/Content/MermaidDiagram.tsx:
```diff
@@ -0,0 +1,327 @@
+        {isRendered && svgContent && (
+          <div
+            className="mermaid-container flex justify-center"
+            dangerouslySetInnerHTML={{ __html: svgContent }}
```
Thanks, TIL! Modified the implementation now to use just `<img` tag our requirements were pretty basic.

ConstantTime commented:
Latest demo after addressing comments:

https://github.com/user-attachments/assets/8d0098aa-031b-40e6-ae68-698418ff8852

ConstantTime commented:
Also created [this cursor rule](https://gist.github.com/ConstantTime/fb0b198db99a94c2f5ff0c5a5d79c870) to make sure this translation change isn't a miss in future PRs. Sharing here for wider reference.

epearson commented:
Does this have the ability to zoom/pan on the svg that is generated by Mermaid like in the Artifacts feature? For larger Gantt and other charts you will have a hard time reading the charts without it.

ConstantTime commented:
> @ConstantTime can you also resolve the zoom/pan on the svg?

Yes @danny-avila, I am going to work on this tonight.

danny-avila commented:
@ConstantTime no worries, thanks for working on this!

ConstantTime commented:
@danny-avila PR up for review fyi.

ConstantTime commented:
Also FYI while I initially explored implementing zoom/pan with pure CSS transforms (similar to the DialogImage component), `react-zoom-pan-pinch` provides significantly better gesture support (based on my testing). This is why I added this new library and bumped up the cache size a bit.

danny-avila commented:
adding sourcemap: true and removing usage of sourcemapExclude, we can see the following better after build:

<html>
<body>
<!--StartFragment--><div aria-expanded="true" tabindex="-1" style="box-sizing: border-box; min-width: 0px; min-height: 0px; flex: 0 0 auto; color: rgb(240, 223, 215); font-family: monospace; font-size: 12px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(92, 65, 49); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><span class="source-code" style="box-sizing: border-box; min-width: 0px; min-height: 0px; font-family: monospace; white-space: pre-wrap; font-size: 12px !important; line-height: 1.2;"><span class="console-message-anchor" style="box-sizing: border-box; min-width: 0px; min-height: 0px; float: right; text-align: right; max-width: 100%; margin-left: 4px;"><button class="devtools-link text-button link-style ignore-list-link" jslog="Link; context: script-location; track: click" role="link" tabindex="-1" title="http://localhost:3080/node_modules/lodash-es/_freeGlobal.js:2" style="box-sizing: border-box; min-width: 0px; min-height: 0px; font: inherit; color: rgb(255, 182, 138); text-decoration: underline; outline-offset: 2px; margin: 0px; height: unset; border: none; border-radius: 2px; padding: 0px !important; background: none; flex: 0 0 auto; white-space: nowrap; cursor: pointer; outline: none; max-width: 100%; overflow: hidden; text-overflow: ellipsis; word-break: break-all;"><br class="Apple-interchange-newline">_freeGlobal.js:2</button> </span><span class="console-message-text" style="box-sizing: border-box; min-width: 0px; min-height: 0px; color: rgb(255, 218, 214) !important;">Uncaught <span class="object-value-error source-code" style="box-sizing: border-box; min-width: 0px; min-height: 0px; font-family: monospace; white-space: pre-wrap; font-size: 12px !important; line-height: 1.2;"><span style="box-sizing: border-box; min-width: 0px; min-height: 0px;">ReferenceError: Cannot access 't' before initialization
<span class="formatted-stack-frame" style="box-sizing: border-box; min-width: 0px; min-height: 0px; display: inline; opacity: 0.6;">    at <button class="devtools-link text-button link-style ignore-list-link" jslog="Link; context: script-location; track: click" role="link" tabindex="-1" title="http://localhost:3080/node_modules/lodash-es/_freeGlobal.js:2:35" style="box-sizing: border-box; min-width: 0px; min-height: 0px; font: inherit; color: rgb(255, 182, 138); text-decoration: underline; outline-offset: 2px; margin: 0px; height: unset; border: none; border-radius: 2px; padding: 0px !important; background: none; flex: 0 0 auto; white-space: nowrap; cursor: pointer; outline: none; max-width: 100%; overflow: hidden; text-overflow: ellipsis; word-break: break-all;">_freeGlobal.js:2:35</button></span></span></span></span></span></div><div class="" role="group" style="box-sizing: border-box; min-width: 0px; min-height: 0px; flex: 0 0 auto; color: rgb(240, 223, 215); font-family: monospace; font-size: 12px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(92, 65, 49); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><span class="monospace stack-preview-container width-constrained" style="display: inline-block; width: 1031.24px; box-sizing: border-box; min-width: 0px; min-height: 0px; font-family: monospace; font-size: 12px !important;">
  | (anonymous) | @ | _freeGlobal.js:2
-- | -- | -- | --


</span></div><!--EndFragment-->
</body>
</html>

this is pointing to `freeGlobal` of lodash-es

```
/** Detect free variable `global` from Node.js. */
var freeGlobal = typeof global == 'object' && global && global.Object === Object && global;

export default freeGlobal;

```

`mermaid` package does introduce `lodash-es` and may be problematic for whatever reason.. Just thought I'd share what I found so far

ConstantTime commented:
Fixed the CI bust:

• Dropped the `lodash-es → lodash` alias in `client/vite.config.ts`; mermaid now gets the proper ESM build, so no more “Cannot access ‘t’ before init” crash.  

Client build is green again and diagrams are also rendering fine.

cc @danny-avila

danny-avila commented:
@ConstantTime still having same issues, the entire client fails to render. It seems to happen whenever `mermaid` is imported, dynamically or not. The underlying code/vite config does not seem to affect this, as a minimal setup using mermaid will cause the issue.

To really test this, you need to build the frontend and not use the hot reload from `npm run frontend:dev`.

Running `npm run reinstall` does this with a clean install. Upon making client changes, `cd client && npm run build` will build the production files to test this behavior, too.

I know we wanted to move away from `sandpack` due to the "heaviness" but it may be the only way for now. I created a dedicated inline mermaid diagram just using sandpack here:

https://github.com/danny-avila/LibreChat/compare/main...feat/mermaid-diag-support


<img width="1223" height="627" alt="chrome_TIyUAPpkis" src="https://github.com/user-attachments/assets/15eb230c-eb64-4858-82d5-87e0ab5cf587" />

if you want to work off that branch for styling, please do so.

I believe we can optimize the sandpack rendering even further by not using react at all and just raw html

ConstantTime commented on .gitignore:
```diff
@@ -125,3 +125,4 @@ helm/**/.values.yaml
 
 # SAML Idp cert
 *.cert
+CLAUDE.md
```
Added since multiple folks will have their own CLAUDE.md version

ConstantTime committed:
feat: add Sandpack-based Mermaid diagram support

ConstantTime commented:
Hey @danny-avila 👋 
gentle ping for this if we are planning to take this live in next release.

rohitpathak21 commented:
This is such an awesome addition! Mermaid support in chat makes so much sense , it’s super helpful for visualizing things like flows, systems, or relationships without needing to leave the conversation. I can already see how useful this will be for devs and technical users who think in diagrams. Really clean work and a super cool feature to have in LibreChat. Great job!

I’ve also been working on something similar, a chart rendering feature that lets agents return structured ECharts JSON, which gets rendered inline in chat. If you get a chance, I’d love for you to check it out: [[PR #8634](https://github.com/danny-avila/LibreChat/pull/8634)](https://github.com/danny-avila/LibreChat/pull/8215). I also DMed you on Discord about it , would really appreciate any feedback or thoughts!

ConstantTime commented:
> Hey @danny-avila 👋 gentle ping for this if we are planning to take this live in next release.

Sorry for the ping again @danny-avila but just want to know if I am doing something wrong here or anything pending on me for you to review.

rohitpathak21 commented:
@ConstantTime 
Hey! did you get time to check out the PR

danny-avila commented:
thanks @ConstantTime please resolve the merge conflict

ristomatti commented:
Thank you for pushing this through @ConstantTime, great work!

Today I was stumped as I was sure inline Mermaid worked at some point. To get it working in the sidebar, I had to add the snippet from documentation as part of the agent instructions to make it work (sometimes). Claude 4 Sonnet was otherwise generating SVG diagrams instead.

I remember there to be a checkbox to enable Mermaid diagrams in the artifact settings, but either it has removed or my memory is badly corrupted... I do not have custom prompt enabled in either the agent or user settings.

P.S. There seems to be conflicts on this PR again.

ConstantTime committed:
Merge branch 'dev' into feat/mermaid_diag_support

ConstantTime commented:
> Thank you for pushing this through @ConstantTime, great work!
> 
> Today I was stumped as I was sure inline Mermaid worked at some point. To get it working in the sidebar, I had to add the snippet from documentation as part of the agent instructions to make it work (sometimes). Claude 4 Sonnet was otherwise generating SVG diagrams instead.
> 
> I remember there to be a checkbox to enable Mermaid diagrams in the artifact settings, but either it has removed or my memory is badly corrupted... I do not have custom prompt enabled in either the agent or user settings.
> 
> P.S. There seems to be conflicts on this PR again.

Hey @ristomatti 
Thanks for taking a look. Addressed the conflicts. Would be super happy if we can now take this to merge :)

ConstantTime commented on .gitignore:
```diff
@@ -134,3 +134,4 @@ helm/**/.values.yaml
 /.openai/
 /.tabnine/
 /.codeium
+CLAUDE.md
```
Oh sorry, I think I somehow missed your comment here. Removed now

@nhtruong Request you to take a look again please.

ConstantTime committed:
fix gitignore

TedLin1993 commented:
@danny-avila is any update? I like this change.
Artifact is too heavy for a simple mermaid
